internal/controller/reinstall/reinstall_test.go

```diff
@@ -666,24 +687,645 @@ var _ = Describe("ProcessRequest", func() {
+		Expect(err).To(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{Requeue: true}))
+
+		cond := findReinstallStatusCondition(clusterInstance, v1alpha1.ReinstallPreservationDataBackedup)
+		Expect(cond).ToNot(BeNil())
+		Expect(cond.Status).To(Equal(metav1.ConditionFalse))
+		Expect(cond.Reason).To(Equal(string(v1alpha1.DataUnavailable)))
+
+		cond = findReinstallStatusCondition(clusterInstance, v1alpha1.ReinstallClusterIdentityDataDetected)
+		Expect(cond).ToNot(BeNil())
+		Expect(cond.Status).To(Equal(metav1.ConditionFalse))
+		Expect(cond.Reason).To(Equal(string(v1alpha1.Failed)))
+	})
+
+})
+
+var _ = Describe("ensurePreservedDataIsRestored", func() {
+	var (
+		ctx        context.Context
+		c          client.Client
+		testLogger *zap.Logger
+		handler    *ReinstallHandler
+
+		testNamespace = "test"
+
+		clusterInstance *v1alpha1.ClusterInstance
+
+		secret1, secret2, secret3, secret4 *corev1.Secret
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		testLogger = zap.NewNop().Named("Test")
+
+		clusterInstance = &v1alpha1.ClusterInstance{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testNamespace,
+				Namespace: testNamespace,
+			},
+			Spec: v1alpha1.ClusterInstanceSpec{
+				Reinstall: &v1alpha1.ReinstallSpec{
+					PreservationMode: v1alpha1.PreservationModeAll,
+				},
+			},
+			Status: v1alpha1.ClusterInstanceStatus{
+				Conditions: []metav1.Condition{},
+				Reinstall:  &v1alpha1.ReinstallStatus{},
+			},
+		}
+
+		// Create preserved secrets
+		secret1 = createTestSecret("secret1", testNamespace, v1alpha1.PreservationModeClusterIdentity, false)
+		secret2 = createTestSecret("secret2", testNamespace, v1alpha1.PreservationModeClusterIdentity, false)
+		secret3 = createTestSecret("secret3", testNamespace, v1alpha1.PreservationModeAll, false)
+		secret4 = createTestSecret("secret4", testNamespace, v1alpha1.PreservationModeAll, false)
+
+		c = fakeclient.NewClientBuilder().
+			WithScheme(scheme.Scheme).
+			WithStatusSubresource(&v1alpha1.ClusterInstance{}).
+			WithObjects(clusterInstance, secret1, secret2, secret3, secret4).
+			Build()
+
+		handler = &ReinstallHandler{
+			Client:          c,
+			DeletionHandler: &deletion.DeletionHandler{Client: c, Logger: testLogger},
+			Logger:          testLogger,
+		}
+	})
+
+	It("should return immediately if PreservationMode is None", func() {
+		clusterInstance.Spec.Reinstall.PreservationMode = v1alpha1.PreservationModeNone
+
+		result, err := handler.ensurePreservedDataIsRestored(ctx, testLogger, clusterInstance)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(result).To(Equal(reconcile.Result{Requeue: true}))
+		cond := findReinstallStatusCondition(clusterInstance, v1alpha1.ReinstallPreservationDataRestored)
+		Expect(cond.Reason).To(Equal(string(v1alpha1.PreservationNotRequired)))
+	})
+
+	It("should complete restoration successfully even if restoration was previously completed", func() {
```

Comment:
In this case do we overwrite objects if they exist? Should we test for that or is that tested (one way or the other) but the underlying function?

Reply:
We update objects if they already exist - this is tested in the `preservation` package.